File: 2019/day-01/day-01.py
import math as m
import pandas as pd

df = pd.read_csv("input.csv", header=None, names=["mass"])

# Part 1
def get_fuel(mass):
    return(m.floor(mass/3)-2)
df["fuel_for_mass"] = df["mass"].apply(lambda x: get_fuel(x))
print("Part 1: ",df["fuel_for_mass"].sum())

# Part 2
def get_fuel_for_fuel(fuel):
    total_fuel = 0
    while True:
        fuel_for_fuel = get_fuel(fuel)
        if fuel_for_fuel > 0:
            total_fuel += fuel_for_fuel
            fuel = fuel_for_fuel
        else:
            break
    return total_fuel
df["fuel_for_fuel"] = df["fuel_for_mass"].apply(lambda x: get_fuel_for_fuel(x))
print("Part 2: ",df["fuel_for_mass"].sum()+df["fuel_for_fuel"].sum())

File: 2019/day-01/day-01.R

library(tidyverse)

input <- read_csv("input.csv",col_names="mass")

get_fuel <- function(mass){
  return(fuel <- floor(mass/3)-2)
}

# Preprocess
input <- input %>%
  mutate(fuel = get_fuel(mass))

# Part 1
fuel_upper_modules <- input %>%
  summarise(fuel_upper = sum(fuel))
print(paste0("Part 1: ",fuel_upper_modules))

# Part 2
fuel_upper_fuel <- 0
for (row in 1:nrow(input)) {
  mass <- input[row, "fuel"] %>% pull()
  # print(paste0("Step: ",mass))
  continue_loop <- TRUE
  while (continue_loop == TRUE) {
    fuel <- get_fuel(mass)
    # print(paste0("Step: ",fuel))
    if (fuel > 0) {
      fuel_upper_fuel <- fuel_upper_fuel + fuel
      mass <- fuel
    } else {
      continue_loop <- FALSE
    }
  }
}
print(paste0("Part 2: ",fuel_upper_modules+fuel_upper_fuel))
